contrib/mesos/pkg/minion/task.go

```diff
@@ -0,0 +1,519 @@
+func notStartedTask(t *task) taskStateFn {
+	log.Infof("starting task process %q with args '%+v'", t.bin, t.args)
+
+	// create command
+	cmd := exec.Command(t.bin, t.args...)
+	if _, err := cmd.StdoutPipe(); err != nil {
+		// fatal error => terminate minion
+		t.tryError(fmt.Errorf("error getting stdout of %v: %v", t.name, err))
+		return taskShouldRestart
+	}
+	stderrLogs, err := cmd.StderrPipe()
+	if err != nil {
+		// fatal error => terminate minion
+		t.tryError(fmt.Errorf("error getting stderr of %v: %v", t.name, err))
+		return taskShouldRestart
+	}
+
+	t.initLogging(stderrLogs)
+	if len(t.env) > 0 {
+		cmd.Env = t.env
+	}
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid: true,
+	}
+
+	// last min check for shouldQuit here
+	select {
+	case <-t.shouldQuit:
+		return taskShouldRestart
+	default:
+	}
+
+	if err := cmd.Start(); err != nil {
+		t.tryError(fmt.Errorf("failed to start task process %q: %v", t.bin, err))
+		return taskShouldRestart
+	}
+	log.Infoln("task started", t.name)
+	t.cmd = &cmdProcess{delegate: cmd}
+	return taskRunning
+}
+
+type exitError interface {
+	error
+
+	// see os.ProcessState.Sys: returned value can be converted to something like syscall.WaitStatus
+	Sys() interface{}
+}
+
+func taskRunning(t *task) taskStateFn {
+	waiter := t.cmd.Wait
+	type waitResult struct {
+		completed *taskCompletion
+		err       error
+	}
+	trySend := func(wr *waitResult) {
+		if wr.err != nil {
+			t.tryError(wr.err)
+		} else if wr.completed != nil {
+			t.tryComplete(wr.completed)
+		}
+	}
+	// listen for normal process completion in a goroutine; don't block because we need to listen for shouldQuit
+	waitCh := make(chan *waitResult, 1)
+	go func() {
+		wr := &waitResult{}
+		defer func() { waitCh <- wr }()
+
+		if err := waiter(); err != nil {
+			if exitError, ok := err.(exitError); ok {
+				if waitStatus, ok := exitError.Sys().(syscall.WaitStatus); ok {
+					wr.completed = &taskCompletion{
+						name: t.name,
+						code: waitStatus.ExitStatus(),
+					}
+					return
+				}
+			}
+			wr.err = fmt.Errorf("task wait ended strangely for %q: %v", t.bin, err)
+		} else {
+			wr.completed = &taskCompletion{name: t.name}
+		}
+	}()
+
+	// wait for the process to complete, or else for a "quit" signal on the task (at which point we'll attempt to kill manually)
+	select {
+	case <-t.shouldQuit:
+		// check for tie
+		select {
+		case wr := <-waitCh:
+			// we got a signal to quit, but we're already finished; attempt best effort delvery
+			trySend(wr)
+		default:
+			// Wait() has not exited yet, kill the process
+			log.Infof("killing %s : %s", t.name, t.bin)
+			pid, err := t.cmd.Kill()
+			if err != nil {
+				t.tryError(fmt.Errorf("failed to kill process: %q pid %d: %v", t.bin, pid, err))
+			}
+		}
+	case wr := <-waitCh:
+		// task has completed before we were told to quit, pass along completion and error information
+		trySend(wr)
+	}
+	return taskShouldRestart
+}
+
```

Comment:
Everything from here downwards is not about the actual task. IMO it would make sense to move it into its own events.go file.